### ISSUE TITLE
exp: condition contacts-v1 on retraction mode with a `<contacts-v1.backtracking>` doc-type token

### ISSUE THREAD

**Issue by timodonnell** [Open-Athena/MarinFold#175]
## Kind

models

- Kind: models

## Question

Does giving the model an explicit **document-type marker for retraction mode** — so it knows at token 0 whether it is generating a document that may take contacts back — recover the contact-prediction accuracy that [#160](https://github.com/Open-Athena/MarinFold/issues/160)'s 50:50 fine-tune lost, and sharpen the retraction behaviour it learned?

## Hypothesis

**#160 trained on a mixture the model cannot condition on.** Both halves of its corpus begin with the identical prefix:

```
backtracking : <contacts-v1> <begin_sequence> <p714>  ...
clean        : <contacts-v1> <begin_sequence> <p1021> ...
```

There is no marker anywhere in the document. Worse, **20.1% of the #159 backtracking documents contain zero retractions**, so a fifth of that half is indistinguishable from clean data in prefix *and* body.

A model in that situation must marginalise over "am I in a mode where I can take this back later?" at every step. Three of #160's measurements are what that predicts:

- it retracts on only **43% of rollouts** — it is *sampling* the mode rather than being told it;
- its retraction enrichment reaches only **52% of the achievable headroom** (1.134x of a 1.26x ceiling), against the corpus's 97%;
- and, isolated by the readout ablation, **emission quality itself regressed −0.0251 R-precision** — before any retraction is honoured.

That last one is the interesting hypothesis. In retraction mode the optimal emission policy is *more speculative* — you can walk a guess back. With no marker, that speculativeness has nowhere to live except in the shared policy, so it leaks into clean generation. **A model that cannot tell which mode it is in has to hedge in both.**

Prediction: adding the marker recovers most of the −0.0251, and raises the retract rate and enrichment when the backtracking marker is supplied.

## Background

Depends on [#158](https://github.com/Open-Athena/MarinFold/issues/158) (`<retract>` + the fold), [#159](https://github.com/Open-Athena/MarinFold/issues/159) (the corpus), [#160](https://github.com/Open-Athena/MarinFold/issues/160) (the first training run and the evidence above).

#160's [ablation comment](https://github.com/Open-Athena/MarinFold/issues/160#issuecomment-5117903678) is the direct motivation: it decomposed the headline −0.0199 into a **−0.0251 emission-level cost** and a **+0.0052 gain from honouring retractions**, showing the mechanism pays for itself while the training run does not. This experiment attacks the −0.0251.

### 1. Format: a new trailing doc-type token

`<contacts-v1.backtracking>`, following the existing `<contacts-v1.sequence_only>` idiom — a variant doc type swapped in as token 0 by `generate.py`, **not** a new statement type.

**The ordering is the whole safety argument.** `all_domain_tokens()` becomes:

```
[native, contacts-and-distances-v1 block, sequence_only, retract, backtracking]
```

appended **last**, so contacts-v1 goes 2847 → 2848 and every pre-existing id is unchanged.

⚠️ **The superset vocabs need the same treatment or they break.** `contacts_and_crops_v1` / `contacts_and_coordinates_v1` build `inherited_tokens()` by filtering `<retract>` out of contacts-v1's list and re-appending it last. A new trailing contacts-v1 token would otherwise land *inside* the inherited block, shoving the whole xyz/crop block up by one and desyncing the two coordinate formats — the exact hazard #158's `inherited_tokens()` docstring warns about. The new token must be added to that exclusion filter and re-appended last, taking the superset 3849 → 3850.

### 2. Corpus: a one-token rewrite, no regeneration

The marker is determined by **which generator produced the document**, not by its content, and every #159 document came from the backtracking engine. So this is a string rewrite over the published mix, not a re-run of the model-in-the-loop job.

**Mark by generator mode, not by content.** The silent 20.1% keep `<contacts-v1.backtracking>`: they teach the honest conditional *"in this mode, sometimes nothing needs retracting"*. Marking by content would instead teach *"this marker guarantees a retraction"* — a different and probably worse target, and one that would make the marker useless as a mode switch.

### 3. Train

Same recipe as #160 (full fine-tune of exp120, lr 3e-4, wd 0.2, bs 128, seq 8192, 1-epoch cosine) on the marked mix, under a 3850-token superset tokenizer. Needs another +1 offline vocab resize of the exp120 init checkpoint — levanter does not resize on warm start (#160 paid for that lesson).

### 4. Eval — this is where the marker earns its keep

With a marker, the retraction on/off comparison becomes a **generation-time** experiment on one checkpoint rather than a readout-time one:

- prompt with `<contacts-v1>` → clean mode,
- prompt with `<contacts-v1.backtracking>` → retraction mode,

on the same 554-protein eval set, scored through #160's `score_backtracking_worker.py` + exp89 `compute_metrics`. That answers directly whether entering retraction mode is worth it — the question #160 could only approach by re-reading fixed rollouts.

## Success criteria

- **Format is append-only** — every exp120 / #160 id unchanged, both superset vocabs still byte-stable in their xyz/crop block. Asserted in tests, not assumed.
- **The marker is obeyed** — retract rate under `<contacts-v1.backtracking>` clearly above the rate under `<contacts-v1>` (#160's unconditioned model sits at 43%).
- **Emission cost recovered** — clean-mode R-precision closes most of #160's −0.0251 against `exp120-base`.
- **Retraction sharpens** — enrichment captures more than #160's 52% of headroom in backtracking mode.
- **Headline** — best-of-both-modes beats `exp120-base` on the #89 benchmark at matched token compute. This is the bar #160 did not clear.

**Comment by timodonnell** [comment]
## ⚠️ Results here are confounded by a corpus bug — see [#159](https://github.com/Open-Athena/MarinFold/issues/159#issuecomment-5132500722)

#159's closing flush emits the remaining ground-truth contacts **in sorted order** (`backtrack_engine.py:315`), and that flush is ~80% of every backtracking document. The corpus's backtracking half is 0.986 sorted in the post-retraction tail against the clean half's 0.500 — and the trained model reproduces it, at **0.991 tail / 83% fully-sorted rollouts** in retraction mode versus 0.500 in clean mode.

Since the inference recipe scores by **voting across 100 rollouts**, a model doing a sorted sweep emits nearly the same set every time and the vote signal collapses. That is consistent with everything measured here: the R-precision regression, enrichment stuck near 1.1x, and the over-retraction.

**What still stands:** the `<retract>` format (#158), the mode marker (#175) — which works exactly as intended, 0 vs 45 retracts/rollout on the same checkpoint — and all the harness/metric work. **What does not:** any conclusion about whether backtracking helps or hurts contact prediction. Those need re-running against a regenerated corpus.

**Comment by timodonnell** [comment]
## Results — the marker works, retraction still doesn't pay for itself

PR: #197 · Writeup: [`experiments/exp175_models_condition_contacts_v1_on_retraction_mode/README.md`](https://github.com/Open-Athena/MarinFold/blob/exp175-backtracking-mode-token/experiments/exp175_models_condition_contacts_v1_on_retraction_mode/README.md) · Series status: [`experiments/STATUS_backtracking.md`](https://github.com/Open-Athena/MarinFold/blob/exp175-backtracking-mode-token/experiments/STATUS_backtracking.md)

**Two runs, not one.** Partway through the first eval we found the corpus bug in #159 — the ground-truth flush emitted contacts in `sorted()` order, ~80% of every backtracking document. The corpus was regenerated with a shuffled flush and the run repeated identically. `v2` is the result; `v1` is reported beside it because the difference between them turns out to be the cleanest measurement in the whole series.

Everything below: 554 proteins, 100 rollouts each, exp82 rollout+vote, exp89 `compute_metrics`, paired per-protein against `exp120-base`.

### 1. The marker is obeyed — completely

| prompt token 0 | retracts / rollout | proteins with ≥1 retraction |
|---|---|---|
| `<contacts-v1>` | **0.12** | 0.9% |
| `<contacts-v1.backtracking>` | **41.96** | 98.2% |

One checkpoint, two behaviours, selected by a single token. #160's unconditioned model *sampled* the mode instead (43% of rollouts). ✅

### 2. The hedging hypothesis held

| arm | Δ R-precision, all-range | long-range |
|---|---|---|
| #160, unconditioned | −0.0199 ±0.0040 | −0.0225 ±0.0058 |
| **clean mode** | **−0.0063 ±0.0042** | −0.0085 ±0.0049 |
| **retraction mode** | **−0.0153 ±0.0043** | −0.0184 ±0.0065 |

#160's single number sits between the two modes it could not distinguish, which is exactly what "a model that can't tell which mode it's in behaves like a mixture" predicts. Telling it recovers **two thirds** of the clean-mode regression.

![mode comparison](https://raw.githubusercontent.com/Open-Athena/MarinFold/a56881ab2d4f10625ca5783c5fff6b6d467d6b36/experiments/exp175_models_condition_contacts_v1_on_retraction_mode/plots_v2/mode_comparison.png)

### 3. The corpus artifact cost accuracy, not discrimination

This is the finding I did not expect. Across the corpus fix, with everything else identical:

| | v1 (published corpus) → v2 (regenerated) |
|---|---|
| retraction-mode Δ R-precision | −0.0414 → **−0.0153** (63% of the gap) |
| clean-mode Δ R-precision | −0.0068 → −0.0063 (nothing) |
| enrichment, fraction of headroom | 43% → 42% (nothing) |
| model's own rollout sortedness | **0.833 → 0.499** (0.500 = null) |

A clean dissociation. The sorted sweep was collapsing the 100-rollout vote — an *accuracy* mechanism — and was never what taught the model to retract. `measure_sortedness.py` measures this on the models' own rollouts, so the fix is confirmed corpus → training → generation, not just in the corpus.

### 4. Retraction is discriminative, at 42% of ceiling

Enrichment **1.101x** [1.08, 1.12], CI excludes 1.0. P(FP | retracted) = 0.888 against a 0.807 base rate. Normalised against the `1/P(FP)` ceiling: **42% of headroom vs the corpus's 97%**.

Timing transferred almost exactly — delay mean 19.6 / median 9 statements against the corpus's 17.9 / 9. The model learned to *reach back*, not just to undo its last move.

![retraction diagnostics](https://raw.githubusercontent.com/Open-Athena/MarinFold/a56881ab2d4f10625ca5783c5fff6b6d467d6b36/experiments/exp175_models_condition_contacts_v1_on_retraction_mode/plots_v2/retraction.png)

### Success criteria, scored

| | |
|---|---|
| Format is append-only | ✅ 0 id mismatches on 3,849 pre-existing tokens |
| The marker is obeyed | ✅ 0.1 vs 42.0 retracts/rollout, one checkpoint |
| Emission cost recovered | ✅ mostly — −0.0199 → −0.0063, but not to zero |
| Retraction sharpens | ❌ 42% of headroom, flat vs #160 |
| **Beats `exp120-base` on #89** | ❌ **−0.0063 clean, −0.0153 retraction** |

Retraction mode also spends 51% more tokens (757 vs 502 per rollout, 4.6% truncated) to land further behind. At any token budget, the best arm of this experiment is the model we started from.

### Model + notebook

```
hf://buckets/open-athena/MarinFold/checkpoints/exp175-cv1-1_5b-mode50-v2-lr3e-4-e1-cos/hf/step-2070
```

[`notebooks/retraction_mode_playground.ipynb`](https://github.com/Open-Athena/MarinFold/blob/exp175-backtracking-mode-token/notebooks/retraction_mode_playground.ipynb) — free Colab T4, no login, toggle the mode token and watch the retractions. Buckets aren't `snapshot_download`-able, so it lists and downloads through the bucket API with `token=False`.

**Not the accuracy frontier** — `exp120-base` beats it in both modes and #166 beats that. It's published because a checkpoint whose behaviour switches on one token is a useful object to have.

### What I think this means for the series

The 42% transfer gap survived a mode marker *and* a corpus regeneration. The corpus reaches 97% because a ground-truth flush tells it which contacts are wrong; the model gets 42% from the same traces. That gap is between "can be shown the answer" and "can tell from the inside", and no supervised variant has moved it.

So I don't think the next lever is another corpus variant. The natural continuation is to let the model retract and score the *outcome* — the RFT setup #98 already has rollouts for — rather than keep demonstrating retractions it can't yet justify.

One cheap thing that would sharpen the story and is still missing: a **100:0 clean-only control** at the same budget. Nothing in this series separates "backtracking hurt" from "another epoch of fine-tuning on ESM-Atlas hurt".

Leaving this open for you to close.
